File: scripts/apply_tmdb.py
#!/usr/bin/env python3
"""
apply_tmdb.py — Fold TMDB ids from tmdb_matches.csv back into all the deliverables.

Run this AFTER tmdb_match.py has produced tmdb_matches.csv. It adds a `tmdb_id`
to:
  - movies.db            (new column `tmdb_id`)
  - export/films.json    (new top-level field `tmdb_id`)
  - export/films_full.csv and export/films_full.ndjson  (new `tmdb_id` column/field)
  - export/reviews/{id}.json  (new `tmdb_id` field, where a review file exists)

It is idempotent — safe to run multiple times (re-run after improving matches).
By default only rows with status `matched` are applied; add --include-low to also
apply `low_confidence` matches.

Usage
-----
    python3 apply_tmdb.py
    python3 apply_tmdb.py --include-low
    python3 apply_tmdb.py --matches tmdb_matches.csv --db movies.db --export-dir export
"""

import argparse, csv, json, os, sqlite3, sys


def load_matches(path, include_low):
    allowed = {"matched"} | ({"low_confidence"} if include_low else set())
    out = {}
    with open(path, newline='', encoding='utf-8') as fh:
        for row in csv.DictReader(fh):
            tid = (row.get("tmdb_id") or "").strip()
            if row.get("status") in allowed and tid:
                try:
                    out[int(row["id"])] = int(tid)
                except ValueError:
                    pass
    return out


def update_db(db_path, mapping):
    if not os.path.exists(db_path):
        print(f"  skip db: {db_path} not found"); return 0
    con = sqlite3.connect(db_path)
    cols = [r[1] for r in con.execute("PRAGMA table_info(reviews)")]
    if "tmdb_id" not in cols:
        con.execute("ALTER TABLE reviews ADD COLUMN tmdb_id INTEGER")
    con.execute("UPDATE reviews SET tmdb_id = NULL")
    n = 0
    for fid, tid in mapping.items():
        con.execute("UPDATE reviews SET tmdb_id = ? WHERE id = ?", (tid, fid))
        n += 1
    con.commit(); con.close()
    return n


def update_films_json(path, mapping):
    if not os.path.exists(path):
        print(f"  skip films.json: {path} not found"); return 0
    films = json.load(open(path, encoding='utf-8'))
    n = 0
    for f in films:
        tid = mapping.get(f["id"])
        f["tmdb_id"] = tid
        if tid is not None:
            n += 1
    json.dump(films, open(path, "w", encoding='utf-8'), ensure_ascii=False)
    return n


def update_csv(path, mapping):
    if not os.path.exists(path):
        print(f"  skip csv: {path} not found"); return 0
    with open(path, newline='', encoding='utf-8') as fh:
        rows = list(csv.reader(fh))
    header = rows[0]
    if "tmdb_id" not in header:
        header.append("tmdb_id")
    ti = header.index("tmdb_id")
    id_idx = rows[0].index("id") if "id" in rows[0] else 0
    n = 0
    out = [header]
    for r in rows[1:]:
        while len(r) < len(header):
            r.append("")
        try:
            tid = mapping.get(int(r[id_idx]))
        except ValueError:
            tid = None
        r[ti] = str(tid) if tid is not None else ""
        if tid is not None:
            n += 1
        out.append(r)
    with open(path, "w", newline='', encoding='utf-8') as fh:
        csv.writer(fh).writerows(out)
    return n


def update_ndjson(path, mapping):
    if not os.path.exists(path):
        print(f"  skip ndjson: {path} not found"); return 0
    n = 0
    lines = []
    with open(path, encoding='utf-8') as fh:
        for line in fh:
            line = line.strip()
            if not line:
                continue
            o = json.loads(line)
            tid = mapping.get(o.get("id"))
            o["tmdb_id"] = tid
            if tid is not None:
                n += 1
            lines.append(json.dumps(o, ensure_ascii=False))
    with open(path, "w", encoding='utf-8') as fh:
        fh.write("\n".join(lines) + "\n")
    return n


def update_reviews(reviews_dir, mapping):
    if not os.path.isdir(reviews_dir):
        print(f"  skip reviews: {reviews_dir} not found"); return 0
    n = 0
    for fid, tid in mapping.items():
        p = os.path.join(reviews_dir, f"{fid}.json")
        if os.path.exists(p):
            o = json.load(open(p, encoding='utf-8'))
            o["tmdb_id"] = tid
            json.dump(o, open(p, "w", encoding='utf-8'), ensure_ascii=False, indent=2)
            n += 1
    return n


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--matches", default="tmdb_matches.csv")
    ap.add_argument("--db", default="movies.db")
    ap.add_argument("--export-dir", default="export")
    ap.add_argument("--include-low", action="store_true",
                    help="also apply low_confidence matches (default: matched only)")
    args = ap.parse_args()

    if not os.path.exists(args.matches):
        sys.exit(f"ERROR: {args.matches} not found. Run tmdb_match.py first.")

    mapping = load_matches(args.matches, args.include_low)
    print(f"Loaded {len(mapping)} TMDB ids "
          f"({'matched + low_confidence' if args.include_low else 'matched only'}).")

    ex = args.export_dir
    print("Applying:")
    print(f"  movies.db          : {update_db(args.db, mapping)} rows tagged")
    print(f"  films.json         : {update_films_json(os.path.join(ex,'films.json'), mapping)} tagged")
    print(f"  films_full.csv     : {update_csv(os.path.join(ex,'films_full.csv'), mapping)} tagged")
    print(f"  films_full.ndjson  : {update_ndjson(os.path.join(ex,'films_full.ndjson'), mapping)} tagged")
    print(f"  reviews/*.json     : {update_reviews(os.path.join(ex,'reviews'), mapping)} tagged")
    print("\nDone. (Re-runnable — running again refreshes from the current matches CSV.)")


if __name__ == "__main__":
    main()

File: scripts/apply_tmdb.mjs
// Phase 6: fold TMDB matches into Supabase Postgres (the canonical store).
//
// Reads tmdb_matches.csv (produced by scripts/tmdb_match.py) and, for every row
// with status `matched` (add --include-low for weak matches too):
//   - updates movies.{tmdb_id, tmdb_title, tmdb_release_date, poster_path,
//     overview, tmdb_enriched_at}
//   - upserts the referenced TMDB genres into `genres` and links them in
//     `movie_genres`.
//
// Uses the SERVICE ROLE key (bypasses RLS) — run locally only.
//   node --env-file=.env scripts/apply_tmdb.mjs
//   node --env-file=.env scripts/apply_tmdb.mjs --matches tmdb_matches.csv --include-low
//
// Idempotent: re-running refreshes movie fields and re-links genres (movie_genres
// upserts are deduped). It does not need the TMDB API key — only the DB key.

import { readFileSync } from 'node:fs';
import { createClient } from '@supabase/supabase-js';

const { SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY } = process.env;
if (!SUPABASE_URL || !SUPABASE_SERVICE_ROLE_KEY) {
  console.error('Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY. Set them in .env, then:');
  console.error('  node --env-file=.env scripts/apply_tmdb.mjs');
  process.exit(1);
}

// TMDB movie genre ids → names (a fixed, small list, so no API call is needed).
const GENRE_MAP = {
  28: 'Action', 12: 'Adventure', 16: 'Animation', 35: 'Comedy', 80: 'Crime',
  99: 'Documentary', 18: 'Drama', 10751: 'Family', 14: 'Fantasy', 36: 'History',
  27: 'Horror', 10402: 'Music', 9648: 'Mystery', 10749: 'Romance',
  878: 'Science Fiction', 10770: 'TV Movie', 53: 'Thriller', 10752: 'War',
  37: 'Western',
};

// ---- args ----
const argv = process.argv.slice(2);
const getArg = (name, def) => {
  const i = argv.indexOf(name);
  return i >= 0 && argv[i + 1] ? argv[i + 1] : def;
};
const matchesPath = getArg('--matches', 'tmdb_matches.csv');
const includeLow = argv.includes('--include-low');

// ---- minimal RFC-4180 CSV parser (overview can contain commas/quotes/newlines) ----
function parseCSV(text) {
  const rows = [];
  let row = [], field = '', inQuotes = false;
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    if (inQuotes) {
      if (c === '"') {
        if (text[i + 1] === '"') { field += '"'; i++; }
        else inQuotes = false;
      } else field += c;
    } else if (c === '"') {
      inQuotes = true;
    } else if (c === ',') {
      row.push(field); field = '';
    } else if (c === '\n' || c === '\r') {
      if (c === '\r' && text[i + 1] === '\n') i++;
      row.push(field); field = '';
      if (row.length > 1 || row[0] !== '') rows.push(row);
      row = [];
    } else field += c;
  }
  if (field !== '' || row.length) { row.push(field); rows.push(row); }
  if (!rows.length) return [];
  const header = rows[0];
  return rows.slice(1).map((r) => Object.fromEntries(header.map((h, i) => [h, r[i] ?? ''])));
}

const nil = (v) => (v == null || v === '' ? null : v);

let text;
try {
  text = readFileSync(matchesPath, 'utf8');
} catch {
  console.error(`ERROR: ${matchesPath} not found. Run scripts/tmdb_match.py first.`);
  process.exit(1);
}

const allowed = new Set(['matched', ...(includeLow ? ['low_confidence'] : [])]);
const records = parseCSV(text)
  .filter((r) => allowed.has(r.status) && nil(r.tmdb_id))
  .map((r) => ({
    id: Number(r.id),
    tmdb_id: Number(r.tmdb_id),
    tmdb_title: nil(r.tmdb_title),
    tmdb_release_date: nil(r.tmdb_release_date),
    poster_path: nil(r.poster_path),
    overview: nil(r.overview),
    genre_ids: (r.genre_ids || '')
      .split(';')
      .map((s) => Number(s.trim()))
      .filter((n) => Number.isFinite(n) && n > 0),
  }))
  .filter((r) => Number.isFinite(r.id) && Number.isFinite(r.tmdb_id));

console.log(`Loaded ${records.length} TMDB matches (${includeLow ? 'matched + low_confidence' : 'matched only'}).`);
if (!records.length) process.exit(0);

const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY, {
  auth: { persistSession: false },
});

// ---- 0) drop any match whose id isn't an actual movie row (films.json has one
//         extra id not present in `movies`) so the FK upserts can't fail a batch ----
const validIds = new Set();
for (let from = 0; ; from += 1000) {
  const { data, error } = await supabase
    .from('movies').select('id').order('id').range(from, from + 999);
  if (error) { console.error('id read failed:', error.message); process.exit(1); }
  for (const r of data) validIds.add(r.id);
  if (data.length < 1000) break;
}
const before = records.length;
const valid = records.filter((r) => validIds.has(r.id));
if (valid.length < before) console.warn(`  ⚠ skipped ${before - valid.length} match(es) with no movies row`);

// ---- 1) upsert the genres that appear, then map tmdb genre id -> genres.id ----
const usedGenreIds = [...new Set(valid.flatMap((r) => r.genre_ids))];
if (usedGenreIds.length) {
  const genreRows = usedGenreIds.map((gid) => ({
    name: GENRE_MAP[gid] || `TMDB ${gid}`,
    tmdb_genre_id: gid,
  }));
  const { error } = await supabase
    .from('genres')
    .upsert(genreRows, { onConflict: 'tmdb_genre_id' });
  if (error) { console.error('genres upsert failed:', error.message); process.exit(1); }
}
const { data: genreRows, error: gErr } = await supabase
  .from('genres')
  .select('id, tmdb_genre_id');
if (gErr) { console.error('genres read failed:', gErr.message); process.exit(1); }
const genreIdByTmdb = new Map(genreRows.map((g) => [g.tmdb_genre_id, g.id]));

// ---- 2) upsert movie_tmdb (1:1 child keyed on movie_id) ----
const now = new Date().toISOString();
const tmdbRows = valid.map((r) => ({
  movie_id: r.id,
  tmdb_id: r.tmdb_id,
  tmdb_title: r.tmdb_title,
  tmdb_release_date: r.tmdb_release_date,
  poster_path: r.poster_path,
  overview: r.overview,
  enriched_at: now,
}));
let enriched = 0;
for (let i = 0; i < tmdbRows.length; i += 500) {
  const batch = tmdbRows.slice(i, i + 500);
  const { error } = await supabase
    .from('movie_tmdb')
    .upsert(batch, { onConflict: 'movie_id' });
  if (error) { console.error('movie_tmdb upsert failed:', error.message); process.exit(1); }
  enriched += batch.length;
  process.stdout.write(`\r  enriched ${enriched}/${tmdbRows.length}`);
}
process.stdout.write('\n');

// ---- 3) link movie_genres (upsert, deduped on the composite PK) ----
const links = valid.flatMap((r) =>
  r.genre_ids
    .map((gid) => genreIdByTmdb.get(gid))
    .filter(Boolean)
    .map((genre_id) => ({ movie_id: r.id, genre_id }))
);
let linked = 0;
for (let i = 0; i < links.length; i += 1000) {
  const batch = links.slice(i, i + 1000);
  const { error } = await supabase
    .from('movie_genres')
    .upsert(batch, { onConflict: 'movie_id,genre_id', ignoreDuplicates: true });
  if (error) { console.error('movie_genres upsert failed:', error.message); process.exit(1); }
  linked += batch.length;
}

console.log(`✓ Done. ${valid.length} movies enriched, ${usedGenreIds.length} genres, ${linked} genre links.`);
